### PR TITLE
Setup the CalicoNotifierProxy early

### PR DESCRIPTION
**Pull request by Lukasa**
Ok, this is a proposed fix for #635. I'm going to try to test this today, but it'll not necessarily be all that easy to test, so @fasaxc may want to run this test as well.

### CODE REVIEW

**Comment by fasaxc** [comment]
@Lukasa I thought that we needed to make the DB connection _after_ we're forked to avoid sharing the socket?

**Comment by Lukasa** [comment]
We do, but this is a terribly named variable: it doesn't contain any DB connections. If you note we only use it with methods that explicitly pass a `context` object, and *that* object is where the DB connections are. So I think we'll be safe here, but I do want us to beat on it for a while to be sure.

(If I'm right, I should enhance this patch to call the variable what it actually is: `plugin`.)

**Comment by fasaxc** [comment]
Gave this a spin on the bobcat rig.  Sadly, it deadlocks immediately.  I think the call to `NeutronManager.get_plugin()` tries to grab the manager lock to create the plugin but the manager lock is already held to do the initialisation (and apparently isn't reentrant).

Our subsequent timer pop to initialise the driver then deadlocks.

**Comment by fasaxc** [comment]
Would it be crazy to monkey-patch that class rather than interposing a proxy?

**Comment by matthewdupre** [comment]
What's the status of this pull request?

**Comment by fasaxc** [comment]
@matthewdupre I tried it out but it didn't work.  Back to @Lukasa for a respin.  I think we need to fix this for the next release, it harms our reliability when we're being used in a low-churn test set-up.  That's the first thing anyone is going to do.

It's also really easy to hit when you're doing manual testing, it just happens that our FV tests tend to create enough VMs to kickstart the plugin before they mess around with SGs.

**Comment by Lukasa** [comment]
No progress since @fasaxc's updates, it's not currently my highest priority.

> Would it be crazy to monkey-patch that class rather than interposing a proxy?

I don't think we can do that. The class needs a reference to a single instance of the mechanism driver, which needs to be the one running in the forked process in question. I don't *think* we can safely assume that the monkeypatched class will survive forking well.

It's *possible* we can come up with a solution where we say there is always only ever one CalicoMechanismDriver in a process, and then have a simple lookup table, but I don't *know* that that's going to be accurate, either now or in the future.

**Comment by Lukasa** [comment]
Yup, this definitely doesn't work. =(

**Comment by Lukasa** [comment]
There is an extra option here: we can actually consume the RPC messages that should be being generated for security group changes.

**Comment by Lukasa** [comment]
Ok, so here's basically the problem as I understand it.

First, our current method of snaffling up security group updates involves attaching a proxy object to the ML2Plugin object. It seems to be extremely unreliable to do this before forking: attempting to do an `eventlet.spawn` call to it doesn't work, presumably because the fork happens while that eventlet is outstanding and the fork causes that eventlet to get blown up (thanks for the suggestion there @neiljerram).

The simplest way to do this *for now* is to do some truly terribly monkeypatching with global state, I think. In some quick tests on my end this appeared to work ok, which is good enough for now, but we should definitely give it a battering.

Longer term, we should think about whether the right thing to do is to actually get onto the end of the RPC queue that Neutron normally uses to handle security group updates. That's the architecturally correct approach, rather than taking the *extremely brittle* method of monkeypatching an object.

**Comment by Lukasa** [comment]
Ok, this patch includes the monkeypatch solution.

I hate it, but a bit of live test suggests it might work. :disappointed_relieved:

@fasaxc, you'll probably want to test with this. Separately, we should aim to find a way to do this properly.

**Comment by Lukasa** [comment]
Note that I've checked that the code is at least in the same place in Icehouse, Juno, and Kilo, so this 'fix' has a reasonable shot at working across all versions.

**Comment by fasaxc** [comment]
Tested this on Ubuntu 14.04 icehouse and it seems to work.  I can restart neutron and then immediately make SG changes, which are picked up.

File: calico/openstack/mech_calico.py [hunk 1]
```diff
@@ -138,6 +150,10 @@ def __init__(self):
         self._periodic_resync_greenlet = None
         self._epoch = 0
 
+        # Tell the monkeypatch where we are.
+        global mech_driver
```
Comment (fasaxc) [inline calico/openstack/mech_calico.py]: Worth asserting that it's None so we catch it if our assumption gets broken?

Comment (Lukasa) [inline calico/openstack/mech_calico.py]: Yeah, why not. Fixed in 8387191fa7f77c5b4991d64efdcd1d23237ccb3c.

**Comment by Lukasa** [comment]
retest this please

**Comment by Lukasa** [comment]
Merging this because @ed-harrison has waved it through. Retroactive code review from @fasaxc would still be valuable.

**Comment by fasaxc** [comment]
LFTM*

* Looks Finessed To Me

**Pull request merged**